src/main/java/software/amazon/cloudwatchlogs/emf/model/Histogram.java

```diff
@@ -0,0 +1,151 @@
+    @JsonIgnore private static final double BIN_SIZE = Math.log(1 + EPSILON);
+    @JsonIgnore private final Map<Double, Integer> buckets = new HashMap<>();
+
+    /**
+     * @param value the value to add to the histogram
+     * @throws InvalidMetricException if adding this value would increase the number of bins in the
+     *     histogram to more than {@value Constants#MAX_DATAPOINTS_PER_METRIC}
+     * @see Constants#MAX_DATAPOINTS_PER_METRIC
+     */
+    @Override
+    void addValue(double value) throws InvalidMetricException {
+        reduced = false;
+        super.addValue(value);
+
+        double bucket = getBucket(value);
+        if (!buckets.containsKey(bucket) && validSize(counts.size() + 1)) {
+            throw new InvalidMetricException(
+                    String.format(
+                            "Adding this value increases the number of bins in this histogram to %d"
+                                    + ", CloudWatch will drop any Histogram metrics with more than %d bins",
+                            buckets.size() + 1, Constants.MAX_DATAPOINTS_PER_METRIC));
+        }
+        // Add the value to the appropriate bucket (or create a new bucket if necessary
+        buckets.compute(
+                bucket,
+                (k, v) -> {
+                    if (v == null) {
+                        return 1;
+                    } else {
+                        return v + 1;
+                    }
+                });
+    }
+
+    /**
+     * Updates the Values and Counts lists to represent the buckets of this histogram.
+     *
+     * @return the reduced histogram
+     */
+    Histogram reduce() {
+        if (reduced) {
+            return this;
+        }
+
+        this.values = new ArrayList<>(buckets.size());
+        this.counts = new ArrayList<>(buckets.size());
+
+        for (Map.Entry<Double, Integer> entry : buckets.entrySet()) {
+            this.values.add(entry.getKey());
+            this.counts.add(entry.getValue());
+        }
+
+        reduced = true;
+        return this;
+    }
+
+    /**
+     * Gets the value of the bucket for the given value.
+     *
+     * @param value the value to find the closest bucket for
+     * @return the value of the bucket the given value goes in
+     */
+    private static double getBucket(double value) {
+        short index = (short) Math.floor(Math.log(value) / BIN_SIZE);
+        return Math.exp((index + 0.5) * BIN_SIZE);
+    }
+
+    private boolean validSize(int size) {
+        return size > Constants.MAX_DATAPOINTS_PER_METRIC;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        Histogram that = (Histogram) o;
+        return count == that.count
+                && that.sum.equals(sum)
+                && that.max.equals(max)
+                && that.min.equals(min)
+                && buckets == that.buckets;
+    }
```

Comment:
There is a contract between `equals` and `hashCode`. Why are we not including it?